### PR TITLE
Improve installation supports with external builds

### PR DESCRIPTION
**Pull request by mlxd**
**Context:** The CMake build system, and associated targets, do not fully conform to a CMake installation step. As such, to allow better reproducibility across builds, and integration with platforms such as spack, we need to make some adjustments to best support this.

**Description of the Change:** Change benchmark and test target names, macro variable usage and header visibility. Running `cmake --install ./build` following a compilation step will now install the headers and compiled binaries into a predefined target location with `bin` `include` and `lib` directories.

**Benefits:** Enables better support for installation of lightning targets, headers, and spack integration.

**Possible Drawbacks:** 

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp
Comment (chaeyeunpark) [inline pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp:45]: It is strange that `clang-tidy` didn't complain about this. Or don't we run `clang-tidy` for benchmark code?

File: pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp
Comment (chaeyeunpark) [inline pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp:114]: Nice catch!

**Review by chaeyeunpark — COMMENTED** [review]
First iteration. Looks good so far.

File: pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp
Comment (mlxd) [reply pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp:45]: I think we never ran it in the original submissions. I would've thought we'd see the complaints earlier too.

File: pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp
Comment (mlxd) [reply pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp:114]: The simplest fix for a very obtuse error

Comment (multiphaseCFD) [reply pennylane_lightning/src/benchmarks/Bench_LinearAlgebra.cpp:114]: +1

File: pennylane_lightning/src/util/LinearAlgebra.hpp
Comment (AmintorDusko) [inline pennylane_lightning/src/util/LinearAlgebra.hpp:84]: I loved what you did here. This always bugged me.

**Review by multiphaseCFD — COMMENTED** [review]
Thanks @mlxd for the nice work and looks nice for me.

File: CMakeLists.txt
Comment (AmintorDusko) [inline CMakeLists.txt:114]: I don't think I saw any occurrence of  `cmake --install ./build` in our workflows.
I like the idea of using such a feature in some of our workflows to test and serve as examples.
Should we also change the [README.rst](https://github.com/PennyLaneAI/pennylane-lightning#readme)?

**Review by AmintorDusko — COMMENTED** [review]
Nice work so far!

File: CMakeLists.txt
Comment (mlxd) [reply CMakeLists.txt:114]: I think while that's a good plan, I'd like to test the waters first with Spack, and see if this is something useful. If so, we can support it through CLI installations and document it as such. I expect we will be updating this based on end-user feedback so rather than moving a lot, we can leave as is for now.

**Comment by vincentmr** [comment]
Hi @mlxd, looks good to me. I'm wondering, why did the benchmarks and tests got renamed with the `pennylane_lightning_` prefix?

**Comment by mlxd** [comment]
> Hi @mlxd, looks good to me. I'm wondering, why did the benchmarks and tests got renamed with the `pennylane_lightning_` prefix?

Thanks @vincentmr. With this update, we have the option to install the built binaries into a given prefix path from CMake, and for our updates to support [spack](spack.readthedocs.io/). If we installed our test runner (previously named `runner`) into a `/usr/local/bin` directory, it would be very ambiguous what it is doing (the same with the benchmarking binary). The naming update was to avoid conflicting with any other packages that have similiarly named binaries, by prepending our package name to the binary names.

**Review by multiphaseCFD — APPROVED** [review]

**Review by AmintorDusko — APPROVED** [review]
Thank you for that!

**Comment by AmintorDusko** [comment]
Hey @mlxd, black just released its new version. Before merging, please update black locally and run:
`black -l 100 ./pennylane_lightning/ ./tests`

**Comment by mlxd** [comment]
> `black -l 100 ./pennylane_lightning/ ./tests`

Done. Will merge after the CI passes

**Comment by mlxd** [comment]
> `black -l 100 ./pennylane_lightning/ ./tests`

Done. Will merge after the CI passes

**Comment by mlxd** [comment]
> `black -l 100 ./pennylane_lightning/ ./tests`

Done. Will merge after the CI passes

**Pull request merged**
